### ISSUE TITLE
pin scikit-learn version to less than / equal to 0.18.2

### ISSUE THREAD

**Issue by NickleDave** [NickleDave/songdkl#29]
scripts in PCB paper use `GMM`, deprecated in version 0.18.2
https://scikit-learn.org/0.19/whats_new.html

@dgmets reports that 
> there are differences in the Likelihoods emitted from GaussianMixture as compared to the previous GMM module.  In particular, the Likelihoods are strongly impacted by the covariance type used.  This didn't used to be the case.  Anyway, I am going to try to run this down...   The current Dkl measures are proportional to the previous ones, but not the same.

in the meantime we can pin to a version before deprecation.
Using 0.18.2 gives a DeperecationWarning that can be annoying when it gets dumped to stdout 150k times, might be worth using a slightly earlier version

**Comment by NickleDave** [comment]
@sthaar this is the issue I was referring to, looks like I have *not* fixed this

**Comment by NickleDave** [comment]
@sthaar @yangzheng-121
I pushed a branch here where I found dependency versions that seemed to match what was used for the original scripts
https://github.com/NickleDave/songdkl/tree/add-pcb-scripts-generate-oracle
If you clone this repo and check out that branch you can create an environment with the environment.yml file.  
(Notice it is **different** from the environment.yml file that's in the root of the current directory!)

```console
git clone https://github.com/NickleDave/songdkl
cd songdkl
git checkout -t origin/add-pcb-scripts-generate-oracle
conda env create -f environment.yml
```

**Comment by sthaar** [comment]
Hi Dave, Just a bit of feedback, not urgent.

I tried the above code, but either I'm doing something wrong, or the environment.yml file in the brach actually IS the same as the master?? I don't see any differences. If I do env list, I get no songdkl and I do get scikit-learn 0.21.3, same as master.

However, if I do pip install songdkl, it overwrites with scikit learn 0.24.2 (also other versions for numpy, mahotas, matplotlib, and scipy)
I manually uninstalled scikit, then pip installed scikit-learn 0.21.3, mahotas-1.4.9 and numpy-1.17.2 (as listen in .yml)
Now it runs without errors or warnings. 

BUT, is scikit-learn 0.21.3 OK? From the above link I understood it should be 0.19.2. If I manually install that it also works fine, without warnings or errors. The version doesn't seem to affect the number of syllables and the dkl value. I tested only on one folder.

Since this seems to work, I'll just continue using this environment. My current environment ('env list') is attached as txt file

[envsita.txt](https://github.com/NickleDave/songdkl/files/9062271/envsita.txt)

btw, I do get errors on manual install of scikit-learn, yet it _does_ install:

`Collecting scikit-learn===0.19.2
  Downloading https://files.pythonhosted.org/packages/eb/f1/8ab06a3ae2f047098a953e43e1714f47ae0db28d2960afcacaa04d671b51/scikit_learn-0.19.2-cp37-cp37m-manylinux1_x86_64.whl (4.9MB)
     |████████████████████████████████| 4.9MB 7.7MB/s 
ERROR: songdkl 0.1.0 has requirement mahotas<2.0.0,>=1.4.11, but you'll have mahotas 1.4.9 which is incompatible.
ERROR: songdkl 0.1.0 has requirement numpy<2.0.0,>=1.19.4, but you'll have numpy 1.17.2 which is incompatible.
ERROR: songdkl 0.1.0 has requirement scikit-learn<0.25.0,>=0.24.0, but you'll have scikit-learn 0.19.2 which is incompatible.
Installing collected packages: scikit-learn
Successfully installed scikit-learn-0.19.2
`

**Comment by NickleDave** [comment]
Hi @sthaar 

Sorry, I wasn't very clear in my instructions.

I think what might be happening is that when you do `pip install` in the environment, `pip` gets the version of `songdkl` of PyPI, and it installs the dependencies specified for that version, "clobbering" the dependencies you install with the environment.yml file

Can you try instead doing an ["editable" installation](https://pip.pypa.io/en/stable/topics/local-project-installs/) of the local code (not from PyPI) using `pip`? Like so:  
```console
git clone https://github.com/NickleDave/songdkl
cd songdkl
git checkout -t origin/add-pcb-scripts-generate-oracle
conda env create -f environment.yml
# activate the environment, that I confusingly gave the same name as the package
conda activate songdkl
pip install -e .
```
The last two lines are the ones I left out in the snippet above.

If you do it this way you should **not** get `pip` overwriting with e.g. scikit-learn 0.24.x

> BUT, is scikit-learn 0.21.3 OK? From the above link I understood it should be 0.19.2.

It needs to be *less than* version *0.18.2* (not 0.*19*.2) because that is where GMM changed

see release notes here:
https://scikit-learn.org/stable/whats_new/v0.18.html#id5
> The old mixture.GMM is deprecated in favor of the new [mixture.GaussianMixture](https://scikit-learn.org/stable/modules/generated/sklearn.mixture.GaussianMixture.html#sklearn.mixture.GaussianMixture). The new class computes the Gaussian mixture faster than before and some of computational problems have been solved. [#6666](https://github.com/scikit-learn/scikit-learn/issues/6666) by [Wei Xue](https://github.com/xuewei4d) and [Thierry Guillemot](https://github.com/tguillemot).

> If I manually install that it also works fine, without warnings or errors. The version doesn't seem to affect the number of syllables and the dkl value. I tested only on one folder.

If you installed from PyPI in both cases and the scikit-learn version is > 0.18.2 then I would expect you to get the same # of syllables and the same dkl value. If you use scikit-learn < 0.18.2 that has the class `mixture.GMM` instead of the newer  `mixture.GaussianMixture` you should get a different dkl value I think. But I'm not sure yet whether it ultimately affects results

Try for example running the original script (included in this branch) and comparing that output with what you get from the `songdkl` package. You can see the use of `GMM` in the original script here:
https://github.com/NickleDave/songdkl/blob/c8938c31cbf090b4cdecb616b07d2a38663c1895/src/pcb-scripts/pcbi.1006437.s006.py#L202

**Comment by sthaar** [comment]
Ok thanks! `pip install -e .` was the missing info for me.

But I get an error that the setup.py is missing:

`pip install -e .
ERROR: File "setup.py" not found. Directory cannot be installed in editable mode: /home/sitath/songdkl
(A "pyproject.toml" file was found, but editable mode currently requires a setup.py based build.)`

Also, I understand that the version should be lower thatn 0.18.2, which is why I got confused in the first place, but the enveronment.yml file downloaded from this branch `add-pcb-scripts-generate-oracle` (using code above) lists `- scikit-learn=0.21.3=py37hd81dba3_0`. Or will that change after `pip install -e .`??

**Comment by NickleDave** [comment]
Hi @sthaar I'm really sorry for confusing you; after spending some time this morning on this I have a better sense of where I left things.

The short version:
* You are right, the env files that *were* in that branch did not have earlier scikit-learn versions
* I have fixed this, and re-written history for the branch
* The easiest thing for you to do will be to re-clone the entire repo from github
* then do the following, to give you an environment where you can run the original scripts from the paper that are in src/pcb-scripts, and compare the output with what you get from running the current `songdkl` in another environment

```
git clone https://github.com/NickleDave/songdkl
cd songdkl
git checkout -t origin/add-pcb-scripts-generate-oracle
conda env create -f environment.yml
conda activate songdkl-python3.5
cd src/pcb-scripts
python Song_D_KL_calc.py folder_with_songs_from_bird_1 folder_with_songs_from_bird_2 number_of_syllable_calsses_bird_1 number_of_syllable_classes_bird_2
```

* re: doing an editable install of the 'src/songdkl` code into the same conda env:
I fought a bunch with various env things but I was not able to find a way yet to directly install the current version of songdkl into the same environment. I might be able to do so later, e.g. by changing the requires-python and dependencies in the pyproject.toml file to match what's in the conda env, building with flit, and then pip installing the .tar.gz into the env.
Doing an editable install has even more complications I will skip describing for now.

**Comment by sthaar** [comment]
Thanks heeps!!
This works :). 

Just to confirm in case someone wants to reproduce: the songdkl-python3.5 environment indeed has the older scikit-learn. It gives much more reasonable dkl values (0.02, 16 syllables) than the master branch (63.8, 19 syls), when comparing the exact same folder of song. (For zebra finch, 16 syllables is still on the high side though, but there are many other things I should try)

tiny change in code if someone wants to reproduce:
first
`cd src/pcb-scripts`
then
`conda env create -f environment.yml`
`conda activate songdkl-python3.5`

running the original scripts:
for calculating number of syllables:
`python pcbi.1006437.s007.py  folder_with_songs_from_bird_1`
for calculating songdkl:
`python pcbi.1006437.s006.py folder_with_songs_from_bird_1 folder_with_songs_from_bird_2`

**Comment by NickleDave** [comment]
Thank you @sthaar --  glad this is working for you now

I just pushed one more commit so that the script file names match what's in the docstrings at the top of each script

so your example would be:
python calc_sylno.py folder_with_songs_from_bird_1
for calculating songdkl:
python Song_D_KL_calc.py folder_with_songs_from_bird_1 folder_with_songs_from_bird_2

**Comment by sthaar** [comment]
update:
after David changed the python version back to 2.7 (as in original published scripts), the scikit version was 0.19. 
I manually installed scikit 0.17.1 and it runs without error. However, I get a lower nr of syllables: 2 instead of the 20, 16 or 18 I got before. Two syllables, might make more sense in this case, but I'm not sure if the other packages are now the same as in the publication. There are many differences if I compare the env files. See zip attached.

[pcb-scripts_versionsStH.zip](https://github.com/NickleDave/songdkl/files/9130966/pcb-scripts_versionsStH.zip)

**Comment by NickleDave** [comment]
Thank you @sthaar for testing and for providing this .env file.  
To be absolutely sure the environments are the same, we would need access to the original machine.  
Next best thing would be if we could get the ground truth for the number of syllables computed by the original script, e.g. if @dgmets has a spreadsheet with those values, and then find some combination of dependencies that replicates those numbers.    
This would be equivalent to treating the numbers in the spreadsheet as a ["derived oracle"](https://en.wikipedia.org/wiki/Test_oracle) as in #28.
If we can't do that, then the next next best thing would be for us to look by eye at the data from the original paper, agree on some number of syllables, and see if we can get an environment where the script that generates those numbers.
